src/uct/tcp/tcp_sockcm_ep.c

```diff
@@ -518,28 +518,24 @@ static int uct_tcp_sockcm_ep_send_skip_event(uct_tcp_sockcm_ep_t *cep)
         return cep->state & UCT_TCP_SOCKCM_EP_DATA_SENT;
     } else {
         ucs_assert(cep->state & UCT_TCP_SOCKCM_EP_ON_CLIENT);
-        return (cep->state & UCT_TCP_SOCKCM_EP_CLIENT_NOTIFY_SENT) ||
-               ((cep->state & UCT_TCP_SOCKCM_EP_DATA_SENT) &&
-                !(cep->state & UCT_TCP_SOCKCM_EP_CLIENT_NOTIFY_CALLED));
+        /* if data already sent or not packed yet, then skip event */
+        return (cep->state & (UCT_TCP_SOCKCM_EP_CLIENT_NOTIFY_SENT |
+                              UCT_TCP_SOCKCM_EP_DATA_SENT)) ||
+               !(cep->state & UCT_TCP_SOCKCM_EP_PRIV_DATA_PACKED);
```

Comment:
do we even need this skip check function? i would just check sent_length==0 in uct_tcp_sockcm_ep_progress_send. maybe move it to function start.

Reply:
UCT/TCP refactoring is not related to this PR, I did the fix according to current design as separate commit just to avoid blocking UCP change

Reply:
indeed. however as a general question, maybe we can remove it, @alinask @evgeny-leksikov WDYT?

Reply:
@yosefe i think it depends on if the case in here is possible - https://github.com/openucx/ucx/pull/6668/files#r624696346

Reply:
@alinask link seems broken for me, what is the case?

Reply:
it's the comment in this pr, the one below - about if it's possible or not to get to the send function with the UCT_TCP_SOCKCM_EP_DISCONNECTING set